Carto panics when failing to stamp a resource with AlwaysHealthy rules

### Bug description:

When Cartographer attempts to stamp a resource with an `AlwaysHealthy` health rule, it may fail to stamp the object. However, it will still evaluate the `AlwaysHealthy` rule, which, when the resource is reconciled for the first time, will be recognized as having transitioned from `Unknown` to `Healthy` (Cartographer assumes the state is always previously Unknown if it has no record of it being stamped out).

This in turn triggers a `panic` when attempting to record the `ResourceHealthyStatusChanged` event because there is no guard for the stamped object used in the event message being nil at: https://github.com/vmware-tanzu/cartographer/blob/main/pkg/realizer/realizer.go#L167

### Steps to reproduce:
Please provide a set of steps to reproduce the behavior (example):
Apply supply chain with a resource that will fail to stamp (e.g. reference a path that doesn't exist in the template), but has an AlwaysHealthy rule.

### Expected behavior:
Definitely: Cartographer does not panic.
Unclear: should a health status transition really have occurred? Is it really correct that we consider the resource healthy even if it's not stamped?

### Actual behavior:
Cartographer panics and crashes.

### Logs, k8s object dumps:
You'll see a fun panic like this in the logs:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x184c8b0]

goroutine 721 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/Users/scoward/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:118 +0x2d9
panic({0x2d2cc80, 0x3c58290})
	/usr/local/Cellar/go/1.19/libexec/src/runtime/panic.go:890 +0x267
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GetAPIVersion(0x0)
	/Users/scoward/go/pkg/mod/k8s.io/apimachinery@v0.25.2/pkg/apis/meta/v1/unstructured/unstructured.go:218 +0x70
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GroupVersionKind(0x0)
	/Users/scoward/go/pkg/mod/k8s.io/apimachinery@v0.25.2/pkg/apis/meta/v1/unstructured/unstructured.go:426 +0x59
github.com/vmware-tanzu/cartographer/pkg/events.ownerEventRecorder.qualifiedResourceName({{0x30f86a8, 0xc000a9c6c0}, {0x3109710, 0xc0009b17c0}, {0x310ddd8, 0xc000528230}, {{0x310c870, 0xc000f51b30}, 0x0}}, 0x0)
	/Users/scoward/workspace/cartographer/pkg/events/owner_recorder.go:109 +0x90
github.com/vmware-tanzu/cartographer/pkg/events.ownerEventRecorder.ResourceEventf({{0x30f86a8, 0xc000a9c6c0}, {0x3109710, 0xc0009b17c0}, {0x310ddd8, 0xc000528230}, {{0x310c870, 0xc000f51b30}, 0x0}}, {0x2f5744a, ...}, ...)
	/Users/scoward/workspace/cartographer/pkg/events/owner_recorder.go:79 +0xd9
github.com/vmware-tanzu/cartographer/pkg/realizer.(*realizer).Realize(0xc000461e20, {0x310a478, 0xc00139fd70}, {0x30f3c40, 0xc0003ed880}, {0xc000663458, 0x12}, {0xc000184100, 0x7, 0x8}, ...)
	/Users/scoward/workspace/cartographer/pkg/realizer/realizer.go:159 +0xe90
github.com/vmware-tanzu/cartographer/pkg/controllers.(*WorkloadReconciler).Reconcile(0xc000b98000, {0x310a478, 0xc0011601b0}, {{{0xc000d48f90, 0x2a}, {0xc000a1f480, 0x10}}})
	/Users/scoward/workspace/cartographer/pkg/controllers/workload_reconciler.go:148 +0x219c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc0007afcc0, {0x310a478, 0xc000f51b00}, {{{0xc000d48f90, 0x2a}, {0xc000a1f480, 0x10}}})
	/Users/scoward/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:121 +0x172
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0007afcc0, {0x310a478, 0xc000f51b00}, {0x2da3000, 0xc000c4f220})
	/Users/scoward/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:320 +0x3d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0007afcc0, {0x310a3d0, 0xc0009b0b80})
	/Users/scoward/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:273 +0x3a5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/Users/scoward/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:234 +0xdc
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/Users/scoward/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:230 +0x4d9
make: *** [run] Error 2
```

### Versions:
 - main branch of carto as of 9bf3708f


### Deployment info:
n/a

### Additional context:
n/a

--- a/pkg/events/owner_recorder.go
+++ b/pkg/events/owner_recorder.go
@@ -76,10 +76,14 @@ func (o ownerEventRecorder) Eventf(eventtype, reason, messageFmt string, args ..
 }
 
 func (o ownerEventRecorder) ResourceEventf(eventtype, reason, messageFmt string, resource *unstructured.Unstructured, args ...interface{}) {
-	qualifiedResourceWithName, err := utils.GetQualifiedResourceWithName(o.mapper, resource)
-	if err != nil {
-		o.log.V(logger.DEBUG).Error(err, "cannot find rest mapping for resource", "resource", resource)
-		return
+	qualifiedResourceWithName := "<nil>"
+	if resource != nil {
+		var mappingErr error
+		qualifiedResourceWithName, mappingErr = utils.GetQualifiedResourceWithName(o.mapper, resource)
+		if mappingErr != nil {
+			o.log.V(logger.DEBUG).Error(mappingErr, "cannot find rest mapping for resource", "resource", resource)
+			return
+		}
 	}
 	messageFmt = strings.ReplaceAll(messageFmt, QualifiedResourceNameToken, qualifiedResourceWithName)
 	o.rec.Eventf(o.obj, eventtype, reason, messageFmt, args...)

--- a/pkg/events/owner_recorder_test.go
+++ b/pkg/events/owner_recorder_test.go
@@ -101,5 +101,17 @@ var _ = Describe("OwnerRecorder", func() {
 				Expect(out).To(Say(`cannot find rest mapping for resource.*"apiVersion":"example.com/v1".*"kind":"foo".*"name":"the-resource"`))
 			})
 		})
+
+		It("handles nil stamped object gracefully", func() {
+			rec.ResourceEventf("Normal", "EggsNil", "Resource [%Q] might panic without a nil check!", nil)
+
+			Expect(fakeRecorder.EventfCallCount()).To(Equal(1))
+			eventObject, eventType, reason, messageFormat, eventMessageArgs := fakeRecorder.EventfArgsForCall(0)
+			Expect(eventObject).To(Equal(ownerObject))
+			Expect(eventType).To(Equal("Normal"))
+			Expect(reason).To(Equal("EggsNil"))
+			Expect(messageFormat).To(Equal(`Resource [<nil>] might panic without a nil check!`))
+			Expect(eventMessageArgs).To(BeEmpty())
+		})
 	})
 })

